bouncer/shared/fill_or_kill.ts

```diff
@@ -0,0 +1,139 @@
+  isWithinOnePercent,
+  newAddress,
+  observeBalanceIncrease,
+  shortChainFromAsset,
+} from './utils';
+import { requestNewSwap } from './perform_swap';
+import { send } from './send';
+import { getBalance } from './get_balance';
+import { getChainflipApi, observeEvent } from './utils/substrate';
+import { RefundParameters } from './new_swap';
+
+/// Do a swap with unrealistic minimum price so it gets retried and then refunded.
+async function testMinPriceRefund(asset: Asset, amount: number) {
+  const swapAsset = asset === Assets.Usdc ? Assets.Flip : Assets.Usdc;
+  const refundAddress = await newAddress(asset, 'FOK_REFUND');
+  const destAddress = await newAddress(swapAsset, 'FOK');
+  console.log(`Swap destination address: ${destAddress}`);
+
+  const refundBalanceBefore = await getBalance(asset, refundAddress);
+
+  const invalidMinPrice = amountToFineAmount(
+    '99999999999999999999999999999999999999999999999999999',
+    assetDecimals(asset),
+  );
+
+  await using chainflip = await getChainflipApi();
+  const currentRetryDuration = Number(await chainflip.query.swapping.swapRetryDelay());
+  console.log(`Current swap retry delay: ${currentRetryDuration} blocks`);
+
+  const refundParameters: RefundParameters = {
+    retryDurationBlocks: 1.6 * currentRetryDuration, // More than the swap retry delay, so it will be retried once.
+    refundAddress:
+      asset === Assets.Dot ? decodeDotAddressForContract(refundAddress) : refundAddress,
+    minPrice: invalidMinPrice,
+  };
+
+  console.log(`Requesting swap from ${asset} to ${swapAsset} with unrealistic min price`);
+  const swapRequest = await requestNewSwap(
+    asset,
+    swapAsset,
+    destAddress,
+    'FoK_Test',
+    undefined, // messageMetadata
+    0, // brokerCommissionBps
+    false, // log
+    0, // boostFeeBps
+    refundParameters,
+  );
+  const depositAddress = swapRequest.depositAddress;
+  const depositChannelId = swapRequest.channelId;
+
+  const observeDestinationAddress =
+    asset === Assets.Dot ? decodeDotAddressForContract(destAddress) : destAddress;
+
+  const observeSwapScheduled = observeEvent('swapping:SwapScheduled', {
+    test: (event) =>
+      typeof event.data.origin === 'object' &&
+      'DepositChannel' in event.data.origin &&
+      event.data.origin.DepositChannel.channelId === depositChannelId.toString() &&
+      event.data.destinationAddress[shortChainFromAsset(swapAsset)]?.toLowerCase() ===
+        observeDestinationAddress.toLowerCase(),
+  }).event;
+
+  // Deposit the asset
+  await send(asset, depositAddress, amount.toString());
+  console.log(`Sent ${amount} ${asset} to ${depositAddress}`);
+
+  const swapScheduledEvent = await observeSwapScheduled;
+  const swapId = Number(swapScheduledEvent.data.swapId.replaceAll(',', ''));
+  console.log(`${asset} swap scheduled, swapId: ${swapId}`);
+
+  // TODO: Observing after the SwapScheduled event means its possible to miss the events, but we need to the swap id.
```

Comment:
Does this mean that this will be a flaky test?

Reply:
Not a problem in practice because we start monitoring right after the swap scheduled event is seen, and the other events are always delayed by a few blocks. But yes, i agree a function that observes with a slight history check would be great.